File: windows/ZegoMediaPlayerVideoManager.cpp
#include "include/zego_express_engine/ZegoMediaPlayerVideoManager.h"

static std::shared_ptr<ZegoMediaPlayerVideoManager> instance_ = nullptr;
static std::once_flag singletonFlag;

void ZegoMediaPlayerVideoManager::onVideoFrame(ZEGO::EXPRESS::IZegoMediaPlayer *mediaPlayer,
                                            const unsigned char **data,
                            unsigned int * dataLength, ZEGO::EXPRESS::ZegoVideoFrameParam param) {
    if (handler_) {
        ZGFlutterVideoFrameParam frameParam;
        frameParam.format = (ZGFlutterVideoFrameFormat)param.format;
        frameParam.height = param.height;
        frameParam.width = param.width;
        frameParam.rotation = param.rotation;
        for (size_t i = 0; i < 4; i++)
        {
            frameParam.strides[i] = param.strides[i];
        }
             
        handler_->onVideoFrame(mediaPlayer->getIndex(), data, dataLength, frameParam);
    }
}

void ZegoMediaPlayerVideoManager::onVideoFrame(ZEGO::EXPRESS::IZegoMediaPlayer *mediaPlayer,
                                                const unsigned char **data,
                            unsigned int * dataLength, ZEGO::EXPRESS::ZegoVideoFrameParam param,
                            const char * extraInfo) {
    if (handler_) {
        ZGFlutterVideoFrameParam frameParam;
        frameParam.format = (ZGFlutterVideoFrameFormat)param.format;
        frameParam.height = param.height;
        frameParam.width = param.width;
        frameParam.rotation = param.rotation;
        for (size_t i = 0; i < 4; i++)
        {
            frameParam.strides[i] = param.strides[i];
        }
             
        handler_->onVideoFrame(mediaPlayer->getIndex(), data, dataLength, frameParam, extraInfo);
    }
}

void ZegoMediaPlayerVideoManager::onBlockBegin(ZEGO::EXPRESS::IZegoMediaPlayer *mediaPlayer,
                    const std::string &path) {
    if (handler_) {
        handler_->onBlockBegin(mediaPlayer->getIndex(), path);
    }
}

unsigned int ZegoMediaPlayerVideoManager::onBlockData(ZEGO::EXPRESS::IZegoMediaPlayer *mediaPlayer,
                            unsigned char *const buffer,
                                    unsigned int bufferSize) {
    if (handler_) {
        return handler_->onBlockData(mediaPlayer->getIndex(), buffer, bufferSize);
    }
    return -1;                                 
}

std::shared_ptr<ZegoMediaPlayerVideoManager> ZegoMediaPlayerVideoManager::getInstance() {
    std::call_once(singletonFlag, [&] {
        instance_ = std::make_shared<ZegoMediaPlayerVideoManager>();
    });
    return instance_;
}

void ZegoMediaPlayerVideoManager::setVideoHandler(std::shared_ptr<IZegoFlutterMediaPlayerVideoHandler> handler) {
    handler_ = handler;
}

File: android/src/main/java/im/zego/zego_express_engine/ZegoMediaPlayerVideoManager.java
package im.zego.zego_express_engine;

import org.json.JSONObject;

import java.nio.ByteBuffer;

import im.zego.zegoexpress.ZegoMediaPlayer;
import im.zego.zegoexpress.callback.IZegoMediaPlayerVideoHandler;
import im.zego.zegoexpress.constants.ZegoVideoMirrorMode;
import im.zego.zegoexpress.entity.ZegoVideoFrameParam;

public class ZegoMediaPlayerVideoManager extends IZegoMediaPlayerVideoHandler {

    private volatile static ZegoMediaPlayerVideoManager singleton;

    private ZegoVideoFrameParam mParam;
    private ZegoVideoMirrorMode mMirrorMode;

    private static IZegoFlutterMediaPlayerVideoHandler mHander;

    public ZegoMediaPlayerVideoManager() {

    }

    /**
     * Get the custom video capture manager instance
     */
    public static synchronized ZegoMediaPlayerVideoManager getInstance() {
        if (singleton == null) {
            singleton = new ZegoMediaPlayerVideoManager();
        }
        return singleton;
    }

    /**
     * Set video data callback handler of the media player.
     *
     * Available since: 2.1.0
     * Description: By setting this callback, the video data of the media resource file played by the media player can be called back.
     * When to call: After the [ZegoMediaPlayer] instance created.
     * Restrictions: None.
     * Caution: When you no longer need to get the video frame data, please call this function again to clear the handler to stop the video frame data callback.
     *
     * @param handler Video data callback handler for media player
     */
    public void setVideoHandler(IZegoFlutterMediaPlayerVideoHandler handler) {
        mHander = handler;
    }

    @Override
    public void onVideoFrame(ZegoMediaPlayer mediaPlayer, ByteBuffer[] data, int[] dataLength,
                             ZegoVideoFrameParam param, JSONObject extraInfo) {
        if (mHander != null) {
            ZGFlutterVideoFrameParam videoFrameParam = new ZGFlutterVideoFrameParam();
            videoFrameParam.width = param.width;
            videoFrameParam.height = param.height;
            videoFrameParam.rotation = param.rotation;
            videoFrameParam.format = ZGFlutterVideoFrameFormat.getVideoFrameFormat(param.format.value());
            for (int i = 0; i < 4; i++) {
                videoFrameParam.strides[i] = param.strides[i];
            }
            mHander.onVideoFrame(mediaPlayer.getIndex(), data, dataLength, videoFrameParam, extraInfo);
        }
    }
}
